modyn/common/trigger_storage_cpp/src/trigger_storage_cpp.cpp

```diff
@@ -0,0 +1,320 @@
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+#include <filesystem>
+#include <fstream>
+#include <future>
+#include <iostream>
+#include <vector>
+
+namespace modyn::common::trigger_storage_cpp {
+
+const int DTYPE_SIZE = 16;
+
+/**
+ * @brief Get the number of samples in file
+ *
+ * @param filename File to read
+ * @return uint64_t Amount of samples
+ */
+uint64_t get_num_samples_in_file_impl(const char* filename) {
+  std::ifstream file = open_file(filename);
+  read_magic(file);
+  return read_data_size_from_header(file);
+}
+
+/**
+ * @brief Get a specified amount of samples from matching files, starting at a specified index
+ *
+ * @param folder Directory to search in
+ * @param size Return pointer for array size
+ * @param pattern Pattern to match. Should match at start of filename
+ * @param start_index Index of the first sample to store
+ * @param worker_subset_size Total amount of samples to store
+ * @return void* Array of samples
+ */
+void* get_worker_samples_impl(const char* folder, uint64_t* size, const char* pattern, const uint64_t start_index,
+                              const std::size_t worker_subset_size) {
+  const std::vector<std::string> matching_files = get_matching_files(folder, pattern);
+  std::size_t samples = 0;
+  std::size_t current_index = 0;
+
+  std::vector<char> char_vector;
+
+  for (const std::string& filename : matching_files) {
+    if (current_index >= start_index + worker_subset_size) {
+      // We have already retrieved all the samples for the worker
+      break;
+    }
+    const std::size_t num_samples = get_num_samples_in_file_impl(filename.c_str());
+    if (current_index + num_samples <= start_index) {
+      // The samples in the file are before the samples for the worker
+      current_index += num_samples;
+      continue;
+    }
+    const std::size_t start = (start_index >= current_index) ? start_index - current_index : 0;
+    if (current_index + num_samples < start_index + worker_subset_size) {
+      // The head of samples for the worker are in the file, either partially from
+      // start_index - current_index to the end of the file if start_index> current_index
+      //  or completely from 0 to the end of the file.
+      // Because the end index is exclusive, we compare < instead of <= otherwise we would retrieve
+      // one more sample than we should
+      parse_file_subset(filename.c_str(), char_vector, samples, start, num_samples);
+      samples += num_samples - start;
+      current_index += num_samples;
+      continue;
+    }
+
+    parse_file_subset(filename.c_str(), char_vector, samples, start, start_index + worker_subset_size - current_index);
+    samples += start_index + worker_subset_size - current_index - start;
+    break;
+  }
+
+  void* data = malloc(sizeof(char) * DTYPE_SIZE * samples);
+
+  memcpy((char*)data, char_vector.data(), sizeof(char) * DTYPE_SIZE * samples);
```

Comment:
I don't like having the data type of the return pointer be uint8_t* because it isn't, it's a mix of int and float. 

added `std::size_t data_length = sizeof(char) * dtype_size * samples;`